### ISSUE TITLE
no error when innodb lock wait timeout should occur

### ISSUE THREAD

**Issue by araker** [sidorares/node-mysql2#699]
Hi, 

I'm debugging a issue where a long running transaction should trigger a lock wait timeout in another query, but instead it fails silently. I've made a simple test case, with as little application specific code as possible. I'm using release 1.5.1. 

![screenshot from 2017-12-19 16 47 32](https://user-images.githubusercontent.com/430230/34165387-6f4a497a-e4dc-11e7-8b61-b73ce26f0d58.png)

The database referenced has two tables, products and models, where models.productId has a foreign_key_constraint to product.productId. 

In the test case, I start the transaction, but never end it (to simulate a long running transaction). I think that in this case the 2nd query should give a error after the wait timeout occurs. I checked my database and the timeout has the default value of 50 seconds. But in the example neither catch statements are reached. 
 
Btw, I had trouble with the indentation, therefore a picture of the code.

**Comment by sidorares** [comment]
you start transaction in one connection and then continue doing queries from (potentially) another.  `pool.getConnection()` returns first available connection which might be different from one where you did `.beginTransaction()`

**Comment by araker** [comment]
Thanks for your response. There are two connections indeed, one for the query with a transaction and one for the query without a transaction. 

I'm not sure if I understand the point you're making, is it that a wait timeout can only occur if the query and the transaction use the same connection? 

I'm not very experienced with transactions, so as far as I understand the following should happen: 

- The transaction locks the row in the product table with id 1500. 
- The 2nd query on the models table tries to reference the product with id 1500, encounters the lock and waits. 
- After 50 seconds the innodb_lock_wait_timeout should activate and give a timeout on the 2nd query. 

Is this a correct assumption, or am I missing something here?

**Comment by sidorares** [comment]
I'm pretty sure transaction causes row level lock (this depends on your db settings and storage engine used). In that case both queries can happen in parallel in two connections. Do you see result returned from insert command? ( `insertId` etc )

**Comment by araker** [comment]
I see I forget to mention a few things, all tables use innodb and I also see the lock on the row with id 1500 in the information_scheme table. The isolation level is repeatable read. I start the transaction first and a second later the insert function.
 
The query from the transaction gives a normal response with insertId, the query from the insert function doesn't give a result. What worries me is that the query from the insert doesn't timeout and that it fails silently.

**Comment by sidorares** [comment]
I'm not sure client side timeouts are implemented currently and 100% same as in mysqljs/mysql driver
Handling this kind of timeouts is a bit tricky, you can't just see error and add another command to the same connection, it has to go only after previous command is finished. So options here are 1) kill connection or 2) kill command from another connection using sql `KILL` query

**Comment by araker** [comment]
I've tested with mysqljs/mysql and using a [client side timeout](https://github.com/mysqljs/mysql#timeouts) on the queries works well. Looking at the code of this library, it doesn't seem to be implemented yet. For now I can implement the timeout in my code, though it would be nice to have this feature in this library as well. Thanks for your advice.
